src/simple-keyring.test.ts

```diff
@@ -32,16 +33,25 @@ const testAccount = {
 
 const notKeyringAddress = '0xbD20F6F5F1616947a39E11926E78ec94817B3931';
 
+const randombytes = jest.requireMock('randombytes');
+jest.mock('randombytes');
+
 describe('simple-keyring', function () {
-  let keyring;
+  let keyring: any; // If you set this to SimpleKeyring instead of any, it opens up a whole can of worms
```

Comment:
Okay, I'll bite 😉 What sort of worms come out?

Reply:
All these errors come out.  Are you thinking we should find a fix for all these errors?

```
    src/simple-keyring.test.ts:101:21 - error TS2339: Property 'isSigned' does not exist on type 'TxData'.

    101       expect(signed.isSigned()).toBe(true);
                            ~~~~~~~~
    src/simple-keyring.test.ts:110:21 - error TS2339: Property 'isSigned' does not exist on type 'TxData'.

    110       expect(signed.isSigned()).toBe(true);
                            ~~~~~~~~
    src/simple-keyring.test.ts:116:44 - error TS2345: Argument of type '""' is not assignable to parameter of type '`0x${string}`'.

    116       await expect(keyring.signTransaction('', tx)).rejects.toThrow(
                                                   ~~
    src/simple-keyring.test.ts:149:21 - error TS2339: Property 'isSigned' does not exist on type 'TxData'.

    149       expect(signed.isSigned()).toBe(false);
                            ~~~~~~~~
    src/simple-keyring.test.ts:208:40 - error TS2345: Argument of type '""' is not assignable to parameter of type '`0x${string}`'.

    208       await expect(keyring.signMessage('', message)).rejects.toThrow(
                                               ~~
    src/simple-keyring.test.ts:266:31 - error TS2345: Argument of type '`0x${string}` | undefined' is not assignable to parameter of type 'string'.
      Type 'undefined' is not assignable to type 'string'.

    266         keyring.removeAccount(addresses[0]);
                                      ~~~~~~~~~~~~
    src/simple-keyring.test.ts:307:48 - error TS2345: Argument of type '""' is not assignable to parameter of type '`0x${string}`'.

    307       await expect(keyring.signPersonalMessage('', message)).rejects.toThrow(
                                                       ~~
    src/simple-keyring.test.ts:495:53 - error TS2345: Argument of type '`0x${string}` | undefined' is not assignable to parameter of type '`0x${string}`'.
      Type 'undefined' is not assignable to type '`0x${string}`'.

    495       const signature = await keyring.signTypedData(address, typedData, {
                                                            ~~~~~~~
    src/simple-keyring.test.ts:568:52 - error TS2345: Argument of type '{}' is not assignable to parameter of type 'Eip1024EncryptedData'.
      Type '{}' is missing the following properties from type 'Eip1024EncryptedData': version, nonce, ephemPublicKey, ciphertext

    568       await expect(keyring.decryptMessage(address, {})).rejects.toThrow(
                                                           ~~
    src/simple-keyring.test.ts:587:9 - error TS2345: Argument of type 'Buffer' is not assignable to parameter of type 'Record<string, string>'.
      Index signature for type 'string' is missing in type 'Buffer'.

    587         privateKey,
                ~~~~~~~~~~
    src/simple-keyring.test.ts:595:40 - error TS2345: Argument of type '""' is not assignable to parameter of type '`0x${string}`'.

    595         keyring.getEncryptionPublicKey('', privateKey),
                                               ~~
    src/simple-keyring.test.ts:602:59 - error TS2345: Argument of type 'Buffer' is not assignable to parameter of type 'Record<string, string>'.

    602         keyring.getEncryptionPublicKey(notKeyringAddress, privateKey),
                                                                  ~~~~~~~~~~
    src/simple-keyring.test.ts:611:49 - error TS2554: Expected 2 arguments, but got 1.

    611       const encryptionPublicKey = await keyring.getEncryptionPublicKey(address);
                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

    src/simple-keyring.ts:136:50
    
    136   async getEncryptionPublicKey(withAccount: Hex, opts: Record<string, string>) {
                                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        An argument for 'opts' was not provided.
    
    src/simple-keyring.test.ts:679:53 - error TS2345: Argument of type '`0x${string}` | undefined' is not assignable to parameter of type '`0x${string}`'.
      Type 'undefined' is not assignable to type '`0x${string}`'.

    679       const signature = await keyring.signTypedData(address, typedData, {

Reply:
I think we ought to, if we can. Otherwise these tests would sort of operate in JavaScript mode instead of TypeScript mode and so we wouldn't be testing the library in the same way that it would end up getting used.

I think we can break these errors down:

* **Argument of type X is not assignable...** — there are three kinds of tests where this shows up:
  1. Tests that intentionally pass bad input in order to exercise various checks. For instance, in the "returns rejected promise if empty address is passed" test, the `signTransaction` method takes a Hex value, but a non-Hex value is being passed. That's fine, and I think in these cases we probably just want to add a `@ts-expect-error` comment to explain what we're doing here.
  2. Tests that access elements of arrays. Our TypeScript config enables [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig#noUncheckedIndexedAccess). This changes how object properties and array elements are accessed such that the result you get will always be `T | undefined` instead of just `T`. The best way to handle these cases might be to `expect` the thing is there before you try to use it.
  3. A couple of tests for `encryptionPublicKey` attempt to pass a Buffer to `keyring.getEncryptionPublicKey`, when it takes a `Record<string, string>`. This seems wrong to me.
* **Property \`isSigned\` does not exist on type 'TxData'** — This is an interesting error to me, but TypeScript is correct — this method doesn't exist on this type. This would seem to indicate that the return type of `signTransaction` is wrong and should probably be corrected.
* **Expected 2 arguments, but got 1** — The second argument to `getEncryptionPublicKey` seems to be optional.

Reply:
Wow, yeah, this was certainly a can of worms, but I think it actually made me improve some things:

- Added the `KeyringOpt` type to stop using `Record` (this is also in the eth-hd-keyring PR)
- `signTransaction` should indeed return a `TypedTransaction` and not a `TxData`
- The `getEncryptionPublicKey` tests that were passing `(address, privateKey)` were actually always passing the wrong parameters, even in the old JavaScript version.  The second parameter is opts and not privateKey.

However, I am still stuck on the following:

- I'm not sure I understand what you said about "The best way to handle these cases might be to `expect` the thing is there before you try to use it."  I initially thought you meant `expect(addresses[0]).toBeDefined()` but that doesn't stop the error.  I can cast it like `keyring.removeAccount(addresses[0] as Hex)`, is that what I should do?